File: src/algorithms/python/Remove_Linked_List_Elements.py
# -*- coding: utf-8 -*-


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    # @param {ListNode} head
    # @param {integer} val
    # @return {ListNode}
    def removeElements(self, head, val):
        pre, node = None, head
        while node:
            if node.val == val:
                if pre:
                    pre.next = node.next
                    node = node.next
                else:
                    head = head.next
                    pre, node = None, head
            else:
                pre = node
                node = node.next
        return head


if __name__ == "__main__":
    solution = Solution()
    nums = [i for i in range(10)]
    nodes = map(ListNode, nums)
    for i in range(1, 10):
        nodes[i - 1].next = nodes[i]
    node = solution.removeElements(nodes[0], 5)
    while node:
        print node.val
        node = node.next

File: src/algorithms/cpp/Remove_Linked_List_Elements.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        ListNode *new_head = new ListNode(-1);
        new_head->next = head;
        ListNode *node = head, *pre = new_head;
        while (node) {
            ListNode *nxt = node->next;
            if (node->val == val) {
                pre->next = nxt;
                delete node;
                node = nxt;
            }
            else {
                pre = node;
                node = nxt;
            }
        }
        return new_head->next;
    }
};

int main() {
    return 0;
}
